### PR TITLE
SS 500 scan and secure

### PR DESCRIPTION
**Pull request by sandstromviktor**
## Description

The task is to make sure that no critical and high vulnerabilities exists using `trivy`. 
In order to fix some of them, i had to update the `dockerfile`. Some vulnerabilities was due to some test modules.
Therefore, I decided to remove the `requirements.txt file` and replace it with `pyproject.toml`.

### Why?
The nice thing with `poetry` and the `pyproject.toml `file is that we can categorize all python packages and install only some of them depending on the situation.  I have created an `extra` field, so when we set an build argument `DISABLE_EXTRAS=true`, then all test stuff is removed from the build. This results in a leaner image (about 400mb compared to 700mb) and no high,critical. 
```bash
[tool.poetry.dependencies]
python = "^3.8.10"

django = "==4.2.7"
django-celery-beat = "==2.5.0"
...etc 

black = { version = "==23.3.0", optional = true }
isort = { version = "==5.12.0", optional = true }
...etc

[tool.poetry.extras]
linters = ["black", "isort", "flake8", "mypy"]
tests = ["pytest-django", "pytest-cov", "hypothesis", "pytest", "moto"]
```
So when we use build-arg `DISABLE_EXTRAS=true` we install without `extras`

## Version upgrades:
I've also updated some versions to fix vulnerabilities
- `django 4.2.5` -> `django 4.2.7` 
- `kubectl 1.28.2` -> `kubectl:1.28.6` 
- `helm 3.12.3 -> helm 3.14`
- Base images `python:3.8-alpine3.17` -> `python:3.8-alpine3.19`

Please, try to deploy this image and see that everything works as expected. It passes CI and E2E, so i hope that it is enough.  
### How to set build-args?
If you run `docker build`, then to just do
```bash
docker build -t <some-tag> --build-arg DISABLE_EXTRAS=true .
```
or you can add it to docker compose by changing 
```
   studio:
    user: "${UID}:${GID}"
    container_name: studio
    build:
      context: .
      args:
          DISABLE_EXTRAS: "true"
```
Note that it is nice to add this to the celery-worker and celery-beat service as well. 

### What's next?
Next, if this works as i think it does, I will add a github action step that builds this "leaner" image for production and staging and the "fatter" for develop. 

## Checklist

_If you're unsure about any of the items below, don't hesitate to ask. We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [X] This pull request is against **develop** branch (not applicable for hotfixes)
- [ ] I have included a link to the issue on GitHub or JIRA (if any)
- [ ] I have included migration files (if there are changes to the model classes)
- [X] I have included, reviewed and executed tests (unit and end2end) to complement my changes
- [ ] I have updated the related documentation (if necessary)
- [ ] I have updated the release notes (releasenotes.md)
- [X] I have added a reviewer for this pull request
- [X] I have added myself as an author for this pull request
- [ ] In the case I have modified settings.py, then I have also updated the studio-settings-configmap.yaml file in serve-charts

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,13 +3,75 @@
 # Note that these settings are not respected with pre-commit run --all-files
 # In that case add configurations to the .pre-commit-config.yaml file.
 
-[project]
+[tool.poetry]
```
Comment (churnikov) [inline pyproject.toml:6]: Mmm, so we are going all in for poetry, huh?
Cool

I was holding it back as an idea for a long while now. But yeah, with optimizing container size it's a good time to start using it.

I'll add some comments regarding it then.

File: Dockerfile [hunk 2]
```diff
@@ -21,23 +25,32 @@ RUN apk add --update --no-cache \
...
-COPY --from=base /usr/local/bin/ /usr/local/bin/
-COPY --from=kubectl /opt/bitnami/kubectl/bin/kubectl /usr/local/bin/
-COPY --from=helm /usr/bin/helm /usr/local/bin/
+# If build-args is set to DISABLE_EXTRA=true, then we skip all superfluous software
+RUN if [ "$DISABLE_EXTRAS" = "true" ]; then \
+    /root/.local/bin/poetry export -f requirements.txt --output requirements.txt; \
+    else /root/.local/bin/poetry export --all-extras  -f requirements.txt --output requirements.txt; \
+    fi
+
+RUN pip install --no-cache-dir -r requirements.txt
```
Comment (churnikov) [inline Dockerfile:44]:
```suggestion
RUN poetry config virtualenvs.create false --global
# If build-args is set to DISABLE_EXTRA=true, then we skip all superfluous software
RUN if [ "$DISABLE_EXTRAS" = "true" ]; then \
    /root/.local/bin/poetry install -n -q --no-cache --only main; \
    else /root/.local/bin/poetry install -n -q --no-cache --all-extras; \
    fi
```

I think `--only main` is the right option

**Review by churnikov — CHANGES_REQUESTED** [review]
Nice one on using poetry.

I would only suggest adding `poetry.lock` file to git. It's worth it for installation speed and reproducibility of the environment.

File: Dockerfile (hunk 2 above)
Comment (sandstromviktor) [reply Dockerfile]: I have fixed this now. The solution was close to this, but not quite.

File: pyproject.toml [hunk 3]
```diff
@@ -3,13 +3,76 @@
...
+repository = "https://github.com/ScilifelabDataCentre/stackn"
 keywords = ["machine learning", "life science", "research", "django", "python"]
 
+[tool.poetry.dependencies]
+python = "^3.8.10"
+
+django = "==4.2.7"
+django-celery-beat = "==2.5.0"
+django-compressor = "==4.4"
+django-cors-headers = "==4.2.0"
```
Comment (churnikov) [inline pyproject.toml:23]: This fixes such kind of error https://stackoverflow.com/a/70371315/8430015
```suggestion
django-cors-headers = "==4.3.1"
```

**Review by churnikov — COMMENTED** [review]
I am getting this error now

```
studio  | Starting the Studio server...
studio  | Watching for file changes with StatReloader
studio  | Exception in thread django-main-thread:
studio  | Traceback (most recent call last):
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
studio  |     self.connect()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 270, in connect
studio  |     self.connection = self.get_new_connection(conn_params)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/postgresql/base.py", line 275, in get_new_connection
studio  |     connection = self.Database.connect(**conn_params)
studio  |   File "/usr/local/lib/python3.8/site-packages/psycopg2/__init__.py", line 122, in connect
studio  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
studio  | psycopg2.OperationalError: could not translate host name "db" to address: Name does not resolve
studio  |
studio  |
studio  | The above exception was the direct cause of the following exception:
studio  |
studio  | Traceback (most recent call last):
studio  |   File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
studio  |     self.run()
studio  |   File "/usr/local/lib/python3.8/threading.py", line 870, in run
studio  |     self._target(*self._args, **self._kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/autoreload.py", line 64, in wrapper
studio  |     fn(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 136, in inner_run
studio  |     self.check_migrations()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 574, in check_migrations
studio  |     executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 18, in __init__
studio  |     self.loader = MigrationLoader(self.connection)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/loader.py", line 58, in __init__
studio  |     self.build_graph()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/loader.py", line 235, in build_graph
studio  |     self.applied_migrations = recorder.applied_migrations()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/recorder.py", line 81, in applied_migrations
studio  |     if self.has_table():
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/recorder.py", line 57, in has_table
studio  |     with self.connection.cursor() as cursor:
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 330, in cursor
studio  |     return self._cursor()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 306, in _cursor
studio  |     self.ensure_connection()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
studio  |     self.connect()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 91, in __exit__
studio  |     raise dj_exc_value.with_traceback(traceback) from exc_value
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
studio  |     self.connect()
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/base/base.py", line 270, in connect
studio  |     self.connection = self.get_new_connection(conn_params)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
studio  |     return func(*args, **kwargs)
studio  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/postgresql/base.py", line 275, in get_new_connection
studio  |     connection = self.Database.connect(**conn_params)
studio  |   File "/usr/local/lib/python3.8/site-packages/psycopg2/__init__.py", line 122, in connect
studio  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
studio  | django.db.utils.OperationalError: could not translate host name "db" to address: Name does not resolve
studio  |
```

**Review by churnikov — APPROVED** [review]

**Review by churnikov — APPROVED** [review]

**Pull request merged**